File: Math/Hard/LET_no902_Numbers At Most N Given Digit Set.py
class Solution:
    def atMostNGivenDigitSet(self, digits: List[str], n: int) -> int:
        target = str(n)
        targetSize, dSize = len(target), len(digits)
        res = 0
        
        for i in range(1, targetSize):
            res += dSize ** i
        
        for i in range(targetSize):
            hasSame = False
            for d in digits:
                if d < target[i]:
                    res += dSize ** (targetSize - i - 1)
                elif d == target[i]:
                    hasSame = True
            if not hasSame:
                return res
        
        return res + 1

File: Math/Hard/LET_no902_Numbers At Most N Given Digit Set.cpp
class Solution {
public:
    int atMostNGivenDigitSet(vector<string>& digits, int n) {
        string target = to_string(n);
        int tSize = target.size(), dSize = digits.size(), res = 0;
        
        for (int i = 1; i < tSize; i++)
            res += pow(dSize, i);
        
        for (int i = 0; i < tSize; i++) {
            bool hasSame = false;
            for (string& d : digits) {
                if (d[0] < target[i])
                    res += pow(dSize, tSize - i - 1);
                else if (d[0] == target[i])
                    hasSame = true;
            }
            if (!hasSame)
                return res;
        }
        
        return res + 1;
    }
};